### ISSUE TITLE
Bug: Removing and re-adding a repo forgets that the name is already registered

### ISSUE THREAD

**Issue by nottrobin** [canonical-ols/build.snapcraft.io#290]
---

Copied from Trello card: https://trello.com/c/KZfeFS7G/334-bug-removing-and-re-adding-a-repo-forgets-that-the-name-is-already-registered
Labels: Bug

**Comment by nottrobin** [comment]
Assets:
https://trello-attachments.s3.amazonaws.com/581b4f456a7f428516c7e24d/58b9a70054a26e57f5b78651/ee37f8dba64076bb71697bce89c33e5b/Screen_Shot_2017-03-03_at_17.21.36.png
https://trello-attachments.s3.amazonaws.com/581b4f456a7f428516c7e24d/58b9a70054a26e57f5b78651/c8ed9b3d10b5f00eb2cc74158c242fe4/Screen_Shot_2017-03-03_at_17.21.42.png

**Comment by evandandrea** [comment]
Add a repo, register the name, remove the repo, and re-add it.

What happens:

- The snap shows up as "not registered" for publishing.
- Attempting to register the already-registered name results in: "Failed to load resource: the server responded with a status of 409 (Conflict)"

What should happen:

- The snap shows as registered for publishing, using the already-registered name.

**Comment by cjwatson** [comment]
Note that the 409 is not in and of itself a problem; it depends on the fine-grained error code, but one of the possibilities for a 409 here is that you already own the name, and BSI handles that case specially.  Of course it's a problem if this leaks through to the user (not counting the curious user poking around with devtools).

**Comment by bartaz** [comment]
@cjwatson @evandandrea  I believe it's caused by the fact that registering the name in BSI does 2 things: registers the name in the store and authorizes the snap to be published to given name.

When the snap is removed and re-added it's a new snap from LP perspective, so even though the name is registered in the store, this snap was not authorized in LP to be published to given name.

@cjwatson Can you confirm this may be the case? Is this something we can fix without changing the way snap are authorized?

**Comment by cjwatson** [comment]
Yes, it's a new snap in Launchpad, but if the user is signed in to the store, then BSI could detect whether a repository's snap name is already owned by the current user in the store, and if so then it could automatically re-authorize the snap.

In the Trello card, I asked that this not be done until #301 is done, otherwise you can get stuck.

**Comment by bartaz** [comment]
@cjwatson Detecting if name is registered in the store (and re-authorizing it) requires BSI to know what is the snap name, meaning this can only be done for repositories that have snapcraft.yaml and the name from it was registered, right?

We cannot re-authorize snap for repository that doesn't have snapcraft.yaml, because we will not know what name was registered to it.

Also, as you mentioned, re-authorizing snap name without UI to change (from #301) it may mean re-added repo will get an old name that user won't be able to change.

**Comment by cjwatson** [comment]
@bartaz That's correct, and I think that's probably OK.

**Comment by bartaz** [comment]
IRC log

```
17:13 <bartaz> cjwatson: I want to make sure my thinking is correct, as I may start working on https://github.com/canonical-ols/build.snapcraft.io/issues/290 tomorrow
17:14 <bartaz> when we re-add a snap/repo (or actually add any new repo, because I believe at this point we can't really tell the difference), assuming it has snapcraft.yaml we can check if name is registered by current user and if so authorize the snap automatically for this name
17:15 <bartaz> but this also requires user to be already signed in to SSO? (for us to have the macaroon?)
17:15 <bartaz> so if they are not signed in yet (new session), we can't automatically authorize the snap?
17:16 ⇐ jamiey quit (jamiey@faun.canonical.com) Quit: My MacBook has gone to sleep. ZZZzzz…
17:16 <cjwatson> bartaz: That's right.  But we could notice this situation as soon as they sign in (which they'd need to on the way to registering the name) and in that case automatically dispatch the relevant actions.
17:18 <bartaz> so, if newly added repo has snapcraft.yaml and user is signed in, we can do everything in the backround
17:19 <cjwatson> Right
17:19 <bartaz> if they are not signed in, repo will appear as Not registered, but as soon as they sign in, we can authorize it automatically - but because we cannot tell the difference between deleted and new repo it will happen for all repos with snapcraft.yaml (if name is already registered)
17:20 <bartaz> and if repo doesn't have snapcraft.yaml - we can't do anything, it will appear as Non registered (because we don't know snap name)
17:21 <cjwatson> We could plug in store_name automatically if we refactored some things to conditionalise on store_name anyway, but it wouldn't really help - we'd still have a "Not authorized" kind of state which would require the same actions to resolve it
```

**Comment by evandandrea** [comment]
That's an odd user experience. Could we not save the authorisations?

**Comment by cjwatson** [comment]
@evandandrea I can't really think how.  Not to mention that it seems deeply peculiar to do so; if you delete something it should be deleted.  (When you asked us to add removals in the first place it was with the reasoning that it should be easy to get back into a first-time state; and in any case I don't think we should be caching authorisations indefinitely for something that users thought they'd deleted.)

**Comment by evandandrea** [comment]
Okay, let's proceed and see if we can't refine the copy in a follow up.

**Comment by bartaz** [comment]
@evandandrea as far as I understand authorizations are per snap, by removing the snap (enabled repo) authorization is removed.

This was part of requirement of removing enabled repos and being able to get back to 'clean account' state.

But I guess @cjwatson can tell more on if and how it can be done differently on LP side.

It's of course a question of user experience, but also on security. Users are granting LP rights to publish snaps in their name. Right now by removing repo they 'revoke' this right.
If we want to 'save authorisations' it means once authorized in LP name/snap will stay authorized even if repo is removed visually from BSI. The name/snap would stay authorized 'forever' just in case they add repo again (?)

Maybe the problem we are facing is that we want to hide 'authorizing' part and only show name registering in the UI. 'Register name' functionality currently does both: registers the name and authorizes LP to publish to it.
Maybe we should be more explicit about it, and for a case where name is already registered, but repo is newly added we should have 'Not authorized' state, where all is needed is for user to go through SSO to authorize newly added snap with already registered name.

**Comment by matthewpaulthomas** [comment]
I think this report is confusing three distinct issues. First:

> Attempting to register the already-registered name results in: "Failed to load resource: the server responded with a status of 409 (Conflict)"

Undoubtedly a bug. when you removed the repo, we told you “The name will remain registered”, so you should be able to resume using it, for this repo or another repo.

Second:

> The snap shows up as "not registered" for publishing.
> …
> What should happen:
> The snap shows as registered for publishing, using the already-registered name.

That’s much more arguable. What are the reasons you might remove a repo?

A. By mistake
B. You’ve changed your mind about ever publishing the snap
C. You have been publishing the snap, but now it’s EOL
D. You have been publishing the snap, but now you want to transfer ownership to someone else
E. You want to continue publishing the snap, but built from a different GitHub repo
F. You want to continue publishing the snap, but not using build.snapcraft.io

Of these reasons, I think remembering the snap name for removed repos would be nice — but not essential — for (A). For all the others, I don’t think it matters whether we remember it or not. If we did remember the name for removed snaps, though, we’d need to make exceptions for (D) and (E) — in those cases, on re-adding the repo, we wouldn’t be able to restore its name because another repo is now using it.

Third:

> &lt;bartaz&gt; when we re-add a snap/repo (or actually add any new repo, because I believe at this point we can't really tell the difference), assuming it has snapcraft.yaml we can check if name is registered by current user and if so authorize the snap automatically for this name

That idea is completely orthogonal to the idea of remembering the registered name. A repo can have a snapcraft.yaml name, a registered name, both, or neither. Auto-registering the snapcraft.yaml name for any newly-added repo would restore the registered name only in the “both” case. It would also result in people more often seeing errors immediately after adding a repo.

I suggest fixing just the first issue, since it blocks undo. The other two require more design.

**Comment by cjwatson** [comment]
On Wed, Mar 08, 2017 at 05:19:30AM -0800, Matthew Thomas wrote:
>      Attempting to register the already-registered name results in: "Failed
>      to load resource: the server responded with a status of 409 (Conflict)"
> 
>    Undoubtedly a bug. when you removed the repo, we told you “The name will
>    remain registered”, so you should be able to resume using it, for this
>    repo or another repo.

To be clear (and sorry if this is repetitive), it's not a bug if the
server expresses the response of "you already own this name" using 409
with "code": "already_owned", and if BSI then interprets that as a
success (since it can distinguish it from other possible 409 responses).
It's only a bug if the non-2xx response in this case is actually visible
in the UI (not counting devtools).

**Comment by bartaz** [comment]
@matthewpaulthomas 

> That idea is completely orthogonal to the idea of remembering the registered name. A repo can have a snapcraft.yaml name, a registered name, both, or neither. 

Yes. From design point of view it's true. My comment was clarification of possible implementation of this. Currently the link between the registered name and the repository is stored in LP as a snap.
When the snap is removed by user (repository removed from the list) we loose any kind of information about this repository, so we simply can't tell if is was registered before. The only source of possible name for such (newly added or re-added) repo is currently snapcraft.yaml file.

But, as you already mentioned. Scenarios for removing and possibly re-adding given repository need more design work.

According to 409 problem: this in a response from register name API that is ONLY visible in developer tools. Application correctly detects this error and interprets it as success (because all it means that user has already registered this name, so we can ignore this error and treat name as registered). So this is not something visible in the UI.

**Comment by magdalena-mirowicz** [comment]
mostly addressed by #444

**Comment by magdalena-mirowicz** [comment]
@jamiedawsonyoung Since you tested 444 can you confirm if that is also fixed? Thanks
